crates/re_query_cache/src/flat_vec_deque.rs

```diff
@@ -0,0 +1,789 @@
+    ///
+    /// Note that the capacity of `self` does not change.
+    ///
+    /// Panics if `entry_index` is out of bounds.
+    #[inline]
+    #[must_use = "use `.truncate()` if you don't need the other half"]
+    pub fn split_off(&mut self, entry_index: usize) -> Self {
+        let value_offset = self.value_offset(entry_index);
+
+        let mut offsets = self.offsets.split_off(entry_index);
+        for offset in &mut offsets {
+            *offset -= value_offset;
+        }
+
+        Self {
+            values: self.values.split_off(value_offset),
+            offsets,
+        }
+    }
+
+    /// Shortens the deque, keeping all entries up to `entry_index` (excluded), and
+    /// dropping the rest.
+    ///
+    /// Panics if `entry_index` is out of bounds.
+    #[inline]
+    pub fn truncate(&mut self, entry_index: usize) {
+        self.offsets.truncate(entry_index);
+        self.values.truncate(self.value_offset(entry_index));
+    }
+
+    /// Removes the entry at `entry_index` from the deque.
+    ///
+    /// This is O(1) if `entry_index` corresponds to either the start or the end of the deque.
+    /// Otherwise, this requires splitting the deque into three pieces, dropping the superfluous
+    /// one, then stitching the two remaining pices back together.
+    ///
+    /// Panics if `entry_index` is out of bounds.
+    pub fn remove(&mut self, entry_index: usize) {
+        let (start_offset, end_offset) = (
+            self.value_offset(entry_index),
+            self.value_offset(entry_index + 1),
+        );
+        let offset_range = end_offset - start_offset;
+
+        if entry_index == self.num_entries() {
+            self.offsets.truncate(self.num_entries() - 1);
+            self.values.truncate(self.values.len() - offset_range);
+            return;
+        } else if entry_index == 0 {
+            *self = self.split_off(entry_index + 1);
+            return;
+        }
+
+        // NOTE: elegant, but way too slow :)
+        // let right = self.split_off(entry_index + 1);
+        // _ = self.split_off(self.num_entries() - 1);
+        // self.push_back_with(right);
+
+        _ = self.offsets.remove(entry_index);
+        for offset in self.offsets.range_mut(entry_index..) {
+            *offset -= offset_range;
+        }
+
+        let right = self.values.split_off(end_offset);
+        self.values.truncate(self.values.len() - offset_range);
+        self.values.extend(right);
+    }
+
+    /// Removes all entries within the given `entry_range` from the deque.
+    ///
+    /// This is O(1) if `entry_range` either starts at the beginning of the deque, or ends at
+    /// the end of the deque, or both.
+    /// Otherwise, this requires splitting the deque into three pieces, dropping the superfluous
+    /// one, then stitching the two remaining pices back together.
+    ///
+    /// Panics if `entry_range` is out of bounds.
+    #[inline]
+    pub fn remove_range(&mut self, entry_range: Range<usize>) {
```

Comment:
I think adding a `assert!(entry_range.start <= entry_range.end);` is a good idea, since the docstring promises panics on out-of-bounds.

Reply:
It can also do an early-out when `entry_range.start == entry_range.end`